src/controller/p2p/preheat/controller.go

```diff
@@ -0,0 +1,168 @@
+	//
+	// params *q.Query : parameters for querying
+	//
+	// If succeed, matched provider instance list will be returned.
+	// Otherwise, a non nil error will be returned
+	//
+	ListInstance(ctx context.Context, query *q.Query) ([]*providerModels.Instance, error)
+
+	// GetInstance the metadata of the specified instance
+	//
+	// id string : ID of the instance being deleted
+	//
+	// If succeed, the metadata with nil error are returned
+	// Otherwise, a non nil error is returned
+	//
+	GetInstance(ctx context.Context, id int64) (*providerModels.Instance, error)
+
+	// Create a new instance for the specified provider.
+	//
+	// If succeed, the ID of the instance will be returned.
+	// Any problems met, a non nil error will be returned.
+	//
+	CreateInstance(ctx context.Context, instance *providerModels.Instance) (int64, error)
+
+	// Delete the specified provider instance.
+	//
+	// id string : ID of the instance being deleted
+	//
+	// Any problems met, a non nil error will be returned.
+	//
+	DeleteInstance(ctx context.Context, id int64) error
+
+	// Update the instance with incremental way;
+	// Including update the enabled flag of the instance.
+	//
+	// id string                     : ID of the instance being updated
+	// properties ...string 				 : The properties being updated
+	//
+	// Any problems met, a non nil error will be returned
+	//
+	UpdateInstance(ctx context.Context, instance *providerModels.Instance, properties ...string) error
+}
+
+var _ Controller = (*controller)(nil)
+
+// controller is the default implementation of Controller interface.
+//
+type controller struct {
+	// For instance
+	iManager instance.Manager
+}
+
+// NewController is constructor of controller
+func NewController() Controller {
+	return &controller{
+		iManager: instance.Mgr,
+	}
+}
+
+// GetAvailableProviders implements @Controller.GetAvailableProviders
+func (cc *controller) GetAvailableProviders() ([]*provider.Metadata, error) {
+	return provider.ListProviders()
+}
+
+// CountInstance implements @Controller.CountInstance
+func (cc *controller) CountInstance(ctx context.Context, query *q.Query) (int64, error) {
+	return cc.iManager.Count(ctx, query)
+}
+
+// List implements @Controller.ListInstance
+func (cc *controller) ListInstance(ctx context.Context, query *q.Query) ([]*providerModels.Instance, error) {
+	return cc.iManager.List(ctx, query)
+}
+
+// CreateInstance implements @Controller.CreateInstance
+func (cc *controller) CreateInstance(ctx context.Context, instance *providerModels.Instance) (int64, error) {
+	if instance == nil {
+		return 0, errors.New("nil instance object provided")
+	}
+
+	// Avoid duplicated endpoint
+	var query = &q.Query{
+		Keywords: map[string]interface{}{
+			"endpoint": instance.Endpoint,
+		},
+	}
+	num, err := cc.iManager.Count(ctx, query)
+	if err != nil {
+		return 0, err
+	}
+	if num > 0 {
+		return 0, ErrorConflict
+	}
+
+	// !WARN: Check healthy status at fronted.
```

Comment:
But don't forget to provide a new health check API endpoint later.

Reply:
Added TODO